lib/1.4/dml-builtins.dml

```diff
@@ -375,13 +378,50 @@ template post_init {
     param _each_post_init = each post_init in (this);
 }
 
+/**
+### after\_domain
+
+The `after_domain` template provides the following non-overridable
+methods:
+
+<dl><dt>
+
+`remove_after()`
+</dt><dd>
+
+Cancel all pending events posted using `after` which
+are associated with the object instantiating the template.
+
+</dd><dt>
+
+`posted_after() -> (bool)`
+</dt><dd>
+
+Returns `true` if there are any pending events posted using `after` which are
+associated with the object instantiating the template.
+</dd></dl>
+
+Both of these methods only concern events posted with an `after` statement
+associated directly with the object instantiating the template, and not
+those associated with any subobject.
+*/
+template after_domain {
+    shared method remove_after() {
+        _cancel_simple_events(dev.obj, cast(this, _identity_t));
+    }
+
+    shared method posted_after() -> (bool) {
+        return _poll_simple_events(dev.obj, cast(this, _identity_t));
+    }
```

Comment:
How's the naming of these methods? They currently mirror the methods of event objects, but I'd prefer `cancel_after()` instead of `remove_after()`. There's also the question if `posted_after()` should even be offered.

Reply:
Also, how's the naming of the template itself? I don't dislike `after_domain` per se, but it kind of relies on leap of logic to understand why it is named like that (that objects instantiating the template are -- in a sense -- a domain for events posted via `after`). If we drop `posted_after()` we can simply name the template `removable_after` or `cancellable_after`, but if we have both methods that naming isn't appropriate.

Reply:
why not inline the template into `object`? Other supertemplates of object are motivated because it's common for shared methods to need its methods, and `object` is a big template to multiple inherit. I don't think this is a strong use case. Making methods part of `object` is conservative, a later step of extracting it into a supertemplate is 100% compatible.

Reply:
I agree cancel_after is a better name.

Regarding existence of posted_after: We should add it iff we can prove that it adds value. A sufficient proof is to find a real use of `event.posted()` that is better expressed as an `after` (i.e., grep for `.posted(`)

Reply:
> why not inline the template into `object`? Other supertemplates of object are motivated because it's common for shared methods to need its methods, and `object` is a big template to multiple inherit. I don't think this is a strong use case. Making methods part of `object` is conservative, a later step of extracting it into a supertemplate is 100% compatible.

I don't see how making methods part of `object` is more conservative. Anyway, yes, I broke it into a template since I figured people would create templates inheriting it and a `reset` template, like this:

```
template hard_reset_after_domain is (hard_reset, after_domain) {
  shared method hard_reset() default {
    hard_reset_subobjs();
    cancel_after();
  }
}
```

Reply:
When I say it's more conservative, I mean that if we create a template today, then we cannot remove it later without breaking compatibility. While if we don't create it today, we can extract it later without breaking compatibility. (apart from the risk for a name clash)

Reply:
Ok, I see.
I personally think we're gonna want a template for this, but there's little harm in inlining it into `object` for the time being. I'll get rid of `posted_after()` at the same time, and rename `remove_after()` to `cancel_after()`

Reply:
Which gives you the natural template name `cancel_after`, unless `posted_after` will resurrect.

If I ever get my perfect hash stuff done, then multiple inheritance will have a zero cost, so it's OK for `hard_reset_after_domain` to inherit `object`. Also, it could happen that we want to unify coroutine domains and after domains, which also can affect naming. So this is a question of how soon we need after_domain.